app/code/community/Bolt/Boltpay/controllers/ShippingController.php

```diff
@@ -168,39 +172,116 @@ public function indexAction()
+        {
+            $request_json = file_get_contents('php://input');
+            $request_data = json_decode($request_json);
 
-        try {
-            $estimate_response = Mage::helper('boltpay/api')->getShippingAndTaxEstimate($quote);
-        } catch (Exception $e) {
-            $estimate_response = null;
+            $address_data = array(
+                'city'          => $request_data->city,
+                'region'        => $request_data->region_code,
+                'region_name'   => $request_data->region_name,
+                'postcode'      => $request_data->zip_code,
+                'country_id'    => $request_data->country_code
+            );
+
+            /** @var Mage_Directory_Model_Country $countryModel */
+            $countryModel = Mage::getModel('directory/country');
+            $countryObj = $countryModel->load($address_data['country_id']);
+            $isRegionAvailable = ($countryObj->getRegionCollection()->getSize() > 0);
+            if (!$isRegionAvailable) {
+                // If country does not have region options for dropdown.
+                $address_data['region'] = $address_data['region_name'];
+            }
+
+            /** @var Mage_Directory_Model_Region $regionModel */
+            $regionModel = Mage::getModel('directory/region');
+            $regionObj= $regionModel->loadByCode($address_data['region'], $address_data['country_id']);
+            $address_data['region_id'] = ($regionObj) ? $regionObj->getId() : null;
+
+            ////////////////////////////////////////////////////////////////////////
+            // Clear previously set estimates.  This helps if this
+            // process fails or is aborted, which will force the actual index action
+            // to get fresh data instead of reading from the session cache
+            ////////////////////////////////////////////////////////////////////////
+            $cacheKeyLocation = $quote->getId() . '_' . $address_data['country_id'];
+            $prefetchCacheKey = $this->generateCacheKey($quote, $address_data);
+            Mage::app()->getCache()->remove('quote_location_' . $cacheKeyLocation);
+            Mage::app()->getCache()->remove($prefetchCacheKey);
+            ////////////////////////////////////////////////////////////////////////
+
+            $quote->getShippingAddress()->addData($address_data);
+            $quote->getBillingAddress()->addData($address_data);
+            try {
+                $estimate_response = Mage::helper('boltpay/api')->getShippingAndTaxEstimate($quote);
+            } catch (Exception $e) {
+                $estimate_response = null;
+            }
+
+            $lifeTime = 600; //10 minutes
+            Mage::app()->getCache()->save(
+                serialize($address_data),
+                'quote_location_' . $cacheKeyLocation,
+                array('BOLT_QUOTE_LOCATION'),
+                $lifeTime
```

Comment:
I assume you want to save `current time + 10min`, not `10min`.

Reply:
I am not sure that i understand you. I set the lifetime of the current cache 10 minutes.

As you previously wrote:
> can you also include timestamp so we won't use stale cache? (let's say it'll expire in 10min)

As i see (maybe i wrong), but lifetime set in seconds and it determine how long cache will be active after creation.
There is no needs to setup it with timestamp + 10minutes i guess.
By default Zend_Cache_Core class set cache lifetime to 3600 as i see.

Reply:
please disregard this comment. I didn't understand how `Mage::app()->getCache()->save(` works